[BUGFIX] Ne pas afficher le message d'alerte de sortie de focus lorsque l'user reprend un nouvel assessment (PIX-4787).

VincentHardouin opened the pull request:
## :unicorn: Problème
Actuellement, sur Pix App, lorsqu'un utilisateur sort d'une épreuve focus, puis quitte l'épreuve et commence/reprend un autre assessment alors il aura encore le message qui lui indique qu'il est sorti de l'épreuve, alors même que l'épreuve peut ne pas être focus.

### Avant
![focus-before](https://user-images.githubusercontent.com/26384707/163626561-4cde8152-0a05-410c-a1cb-77fbd00da2b5.gif)

### Après
![focus-after](https://user-images.githubusercontent.com/26384707/163626575-972d29aa-8c52-444c-91e5-7232a3796eb1.gif)

## :robot: Solution
Les controllers d'Ember sont des singletons, c'est-à-dire que lorsqu'on accède à plusieurs assessments on utilise en réalité la même instance du controller auquel on passe juste des paramètres différents. 

Nous utilisions une propriété trackée qui dépendait du model puis que nous la redéfinissons par un booléen.

Au début du controller :
```
@tracked hasFocusedOutOfWindow = this.model.assessment.hasFocusedOutChallenge;
```

Puis lors de la réassignation : 
```
hasFocusedOutOfWindow = false;
```

Comme il s'agit d'un singleton, lorsqu'on réutilise le controller, nous ne regardons pas la propriété du model comme la propriété est juste égale à un booléen. 
Il fallait alors utiliser un `get`  car les attributs du model sont trackés + le `get` est appelé de nouveau à chaque affichage. Nous obtenons alors : 
```js
 get hasFocusedOutOfWindow() {
   return this.model.assessment.hasFocusedOutChallenge;
 }
```
Grâce à ça dès que nous changeons le modèle le getter fait bien son boulot en mettant à jour l'affichage, ce qui fait que nous n'avons plus besoin de changer la valeur de `hasFocusedOutOfWindow`.

D'autre part, nous ne lui indiquons pas que certains champs doivent être remis à 0. Pour faire cela, il existe la méthode `resetController` qui est déclenchée lorsque la route est quittée ou change (cf : [doc resetController](https://api.emberjs.com/ember/3.28/classes/Route/methods/resetController?anchor=resetController)).  Nous utilisons alors cette méthode pour gérer ce cas. 

## :rainbow: Remarques


## :100: Pour tester
- Se connecter à Pix APP
- Passer une compétence jusqu'à avoir une épreuve focus
- Sortir de la fenêtre de l'épreuve pour perdre le focus
- Quitter l'épreuve à l'aide du bouton en haut à droite
- Reprendre une autre compétence et constater qu'il n'y a pas le message de sortie d'épreuve.

VincentHardouin committed:
fix(mon-pix): add resetcontroller to handle focusout

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4330.review.pix.fr
- App (.org): https://app-pr4330.review.pix.org
- Orga: https://orga-pr4330.review.pix.fr
- Certif: https://certif-pr4330.review.pix.fr
- Admin: https://admin-pr4330.review.pix.fr
- API: https://api-pr4330.review.pix.fr/api/

 Please check it out!

alexandrecoin reviewed (dismissed):
Nous avons décelé une régression.
Pour reproduire : 
- Aller sur une épreuve focus
- Changer d'onglet et revenir
- Constater le message de sortie d'épreuve
- Quitter le challenge et revenir sur le même assessment
- Le message de sortie d'épreuve n'est plus visible

VincentHardouin committed:
fix(mon-pix): remove model based tracked property

hasFocusedOutWindow is a tracked property that corresponds to this.model.assessment.hasFocusedOutChallenge.
When the user leaves the window, this property is manually set to false, leading to a non-use of the model mentionned beforehand.
hasFocusedOutOfWindow can not be set back to its original state through the resetController method as it depends on the model.
We thus create a getter depending on the model that gets updated thanks to Ember reactivity system.

Co-authored-by: Alexandre Coin <alexandre.coin@pix.fr>
Co-authored-by: Matthias Ferraina <matthias

yannbertrand reviewed (approved):
LGTM
